numpy/lib/arraysetops.py

```diff
@@ -324,7 +325,14 @@ def _unique1d(ar, return_index=False, return_inverse=False,
         aux = ar
     mask = np.empty(aux.shape, dtype=np.bool_)
     mask[:1] = True
-    mask[1:] = aux[1:] != aux[:-1]
+    floats = (float, np.float16, np.float32, np.float64)
+    if aux.shape[0] > 0 and isinstance(aux[-1], floats) and np.isnan(aux[-1]):
+        aux_firstnan = np.searchsorted(aux, np.nan, side='left')
```

Comment:
This would ensure that the proposed change work for all dtypes with `nan`-esque: 
`floating`, `complexfloating`, `timedelta64` and `datetime64`.
```suggestion
    if aux.shape[0] > 0 and aux.dtype.kind in "cfmM" and np.isnan(aux[-1]):
        # Ensure that `NaT` is used for time-like dtypes
        nan = np.array(np.nan, dtype=aux.dtype)
        aux_firstnan = np.searchsorted(aux, nan, side='left')
```

Reply:
Continuing from the previous comment: Adding a test for aforementioned dtypes would be wise.